### PR TITLE
update `future.{write,read}` ABIs

### PR DESCRIPTION
**Pull request by dicej**
This implements https://github.com/WebAssembly/component-model/pull/524 insofar as it affects validation, componentization, etc.

- `future.write` now accepts its payload value as up to 4 flat parameters before spilling to linear memory.

- `future.read` takes no payload pointer when it has no payload type

- `{stream,future}.close-{readable,writable}` have been renamed to `{stream,future}.drop-{readable,writable}`

### CODE REVIEW

**Comment by dicej** [comment]
CI failures appear to be due to me using the latest Rust release; will downgrade to the MSRV locally and fix the issues.

File: crates/wasmparser/src/validator/component.rs [hunk 1]
```diff
@@ -1776,19 +1783,50 @@ impl ComponentState {
...
+            types,
+            &CanonicalOptions {
+                string_encoding: StringEncoding::Utf8,
+                memory: options.memory,
+                realloc: None,
+                post_return: None,
+                concurrency: Concurrency::Async { callback: None },
+                core_type: None,
+                gc: false,
+            },
```
Comment (alexcrichton) [inline crates/wasmparser/src/validator/component.rs]: Instead of creating a new `CanonicalOptions` could this `clone()` and manipulate the `options` above? I mostly want to keep things in sync in terms of gc/string encoding/etc for future compatibliity

**Review by alexcrichton — APPROVED** [review]
Looks good 👍 

In the `*.wast` tests I think I only saw renamings, so could you add tests for futures for both reads/writes to ensure properties such as `future.read` has no pointer with no payload and `future.write` uses flat lowerings or a pointer if the flat lowering is too big? Additionally I think it'd be good to have a test that `future.write` doesn't require memory if the lowering fits

File: crates/wasmparser/src/validator/component.rs (hunk 1 above)
Comment (dicej) [reply crates/wasmparser/src/validator/component.rs]: I was thinking the same thing, but then check with Luke, who confirmed that we use `MAX_FLAT_ASYNC_PARAMS` for `future.write` _regardless_ of whether the `future.write` is called sync or async.  I guess I could still clone, but then force the `concurrence` to async.

Comment (alexcrichton) [reply crates/wasmparser/src/validator/component.rs]: Oh right yeah I think this would still overwrite some options, but I want to be sure to preserve options that aren't otherwise overwritten for the purposes of the internal API calls here (e.g. the canonical ABI for GC strings depends on the string encoding which isn't preserved here, nor is GC)

**Comment by dicej** [comment]
> In the `*.wast` tests I think I only saw renamings, so could you add tests for futures for both reads/writes to ensure properties such as `future.read` has no pointer with no payload and `future.write` uses flat lowerings or a pointer if the flat lowering is too big? Additionally I think it'd be good to have a test that `future.write` doesn't require memory if the lowering fits

FWIW, I've tested those cases end-to-end in wasip3-prototyping, plus the fuzz tester is hitting those cases also.  I agree we should have some handwritten WAST tests as well; I'll add those.

**Comment by alexcrichton** [comment]
Yeah all the code here looks correct so I don't expect any failures, mostly just handwritten tests for the well-known edge cases are good to have to catch any possible issues in future refactorings earlier

File: crates/wasmparser/src/validator/component.rs (hunk 1 above)
Comment (dicej) [reply crates/wasmparser/src/validator/component.rs]: The other wrinkle is that `ComponentFuncType::lower` may return a `LoweredFuncType::Existing(_)` rather than `LoweredFuncType::New(_)` when `gc` is `true`, and I couldn't figure out how to extract the core parameter list from the former.  Do you have any tips?

Comment (alexcrichton) [reply crates/wasmparser/src/validator/component.rs]: Nah that's fine to leave as a panic for now, it's more the silent/subtle bugs of switching options by accident I'm worried about. Overall this is a really minor concern so if you'd prefer it's ok to leave as-is.
